stdlib/set.gr

```diff
@@ -0,0 +1,305 @@
+        buckets[idx] := Some({ key, next: bucket });
+        set.size = incr(set.size);
+      };
+    }
+  }
+  // Resize if there are more than 2x the amount of nodes as buckets
+  if (set.size > (Array.length(buckets) * 2)) {
+    resize(set);
+  } else {
+    void
+  }
+}
+
+let rec nodeInBucket = (key, node) => {
+  if (key == node.key) {
+    true
+  } else {
+    match (node.next) {
+      None => false,
+      Some(next) => nodeInBucket(key, next)
+    }
+  }
+}
+
+export let contains = (key, set) => {
+  let buckets = set.buckets;
+  let idx = getBucketIndex(key, buckets);
+  let bucket = buckets[idx];
+  match (bucket) {
+    None => false,
+    Some(node) => nodeInBucket(key, node)
+  }
+}
+
+let rec removeInBucket = (key, node) => {
+  match (node.next) {
+    None => false,
+    Some(next) => {
+      if (key == next.key) {
+        node.next = next.next;
+        true
+      } else {
+        removeInBucket(key, next)
+      }
+    }
+  }
+}
+
+export let remove = (key, set) => {
+  let buckets = set.buckets;
+  let idx = getBucketIndex(key, buckets);
+  let bucket = buckets[idx];
+  match (bucket) {
+    None => void,
+    Some(node) => {
+      // If it is a top-level node, just replace with next node
+      if (key == node.key) {
+        set.size = decr(set.size);
+        buckets[idx] := node.next;
+      } else {
+        if (removeInBucket(key, node)) {
+          set.size = decr(set.size);
+        }
+      }
+    }
+  }
+}
+
+export let size = (set) => {
+  set.size
+}
+
+export let isEmpty = (set) => {
+  size(set) == 0
+}
+
+export let clear = (set) => {
+  set.size = 0;
+  let buckets = set.buckets;
+  Array.forEachi((bucket, idx) => {
+    buckets[idx] := None;
+  }, buckets);
+}
+
+let rec forEachBucket = (fn, node) => {
+  match (node) {
+    None => void,
+    Some({ key, next }) => {
+      fn(key);
+      forEachBucket(fn, next);
+    }
+  }
+}
+
+export let forEach = (fn, set) => {
+  let buckets = set.buckets;
+  Array.forEach((bucket) => {
+    forEachBucket(fn, bucket)
+  }, buckets);
+}
+
+let rec reduceEachBucket = (fn, node, acc) => {
+  match (node) {
+    None => acc,
+    Some({ key, next }) =>
+      reduceEachBucket(fn, next, fn(acc, key))
+  }
+}
+
+export let reduce = (fn, init, set) => {
+  let buckets = set.buckets;
+  let mut acc = init;
+  Array.forEach((bucket) => {
+    acc = reduceEachBucket(fn, bucket, acc)
+  }, buckets);
+  acc
+}
+
+export let filter = (predicate, set) => {
+  let keysToRemove = reduce((list, key) =>
+    if (!predicate(key)) {
+      [key, ...list]
+    } else {
+      list
+    }, [], set);
+  List.forEach((key) => {
+    remove(key, set);
+  }, keysToRemove);
+}
+
+export let reject = (predicate, set) => {
+  filter((key) => !predicate(key), set)
+}
+
+export let toList = (set) => {
+  reduce((list, key) => [key, ...list], [], set)
+}
+
+export let fromList = (list) => {
+  let set = make();
+  List.forEach((key) => {
+    add(key, set);
+  }, List.unique(list));
```

Comment:
We should be able to drop the `unique` call here, since `add` should take care of it (and it'll be faster!)

Reply:
Yes, `add` should be enough. Thank you.